IRGen: the runtime is compacted into the stdlib

compnerd opened the pull request:
Correct the DLL Storage for runtime functions in the standard library
when building for Windows.  This corrects the non-ARC entry points.
Partially resolves SR-7107.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
CC: @aschwaighofer @rjmccall @gottesmm

compnerd commented:
@rjmccall - no, its really not okay to really ignore it, but I dont have a good way to identify the module being emitted in the IR pass.  This is strictly an improvement over what happens currently, but, it does need further improvements.  I was wondering if there was a preferred way to handle this.

rjmccall commented:
I'm not a huge fan of "this fixes half the problem but I don't know how to solve the second half" patches: all too often, the complete solution doesn't really build from the incomplete one, so we're just left with a bunch of cruft.  If the intermediate position is useful on its own, the calculation changes a bit, but if it's just an abstract milestone, let's figure out how we'll fix everything, and then we can talk about how to break the work down into patches.

compnerd commented:
@rjmccall that sounds perfectly reasonable to me.  I wanted to get this put up here so that we can actually discuss this with some additional context.  I had no intention of moving forward with getting this merged as is.   Would you prefer that this discussion be moved to the forums?

I think that the crux of the issue is that we do not have a good way to identify if we are building the standard library.  Even where we can query the information, we rely on the *name* of the module.  Is it acceptable to propagate that hack and just use the name of the LLVM module?  I think that cleaning up the identification of the standard library (formalizing passing `-parse-stdlib` around) should be a separate larger change.

rjmccall commented:
Opening a PR while you're working through an implementation is just fine.  You should say something explicit in the PR about that, though, especially if you're going to tag people to be reviewers. I'm not going to try to infer it just because you've left placeholder text in the PR summary. :)

`ModuleFile` has an `isStdlibModule()` method that ought to be authoritative.  If that's ever insufficient for some reason, well, we can make it sufficient.

I don't think we should use the name of the LLVM module.  IR-level metadata is probably the correct way of passing information to passes.

compnerd commented:
So, the problem is that `ModuleFile` is not accessible when running the module pass.  We could introduce a new LLVM module level metadata for identifying that it is the swift standard library.  I'll see if I can hook that up.

rjmccall commented:
Sounds good.

compnerd committed:
IRGen: add swift module metadata into the LLVM module

Introduce new named metadata in the LLVM module (swift.module.flags)
that is used to identify if the module is the standard library.  This
will be used to correct the DLL Storage for runtime functions.

compnerd committed:
IRGen: special case the handling for the stdlib

Because the runtime is compacted into the standard library, functions
which are normally imported are actually local definitions.  Use module
level named metadata to identify the module as being the swift standard
library.  Refactor the condition slightly to improve code readability.

This addresses SR-7107!

compnerd commented:
@rjmccall okay, I believe that this should be correct now.

compnerd commented:
Ugh, a clean build found a case that wasn't considered: the symbols that are imported via `@_silgen_name` get the wrong storage :/

compnerd commented:
So, this is doing much better, but it doesn't seem to catch everything.  There are a number of Value Witness Tables, Protocol Descriptors, and Type Metadata which are given the wrong storage.    Additionally there are some support functions that are given the wrong storage.

compnerd commented:
@rjmccall @parkera @jckarter - so, at this point, it is only the VWT, Protocol Descriptors, and Type Metadata that is causing an issue.  I am uncertain what the best way to handle this is.  Handle those specifically for the stdlib?  They all go through `getIRLinkage` and `LinkInfo::get` to calculate the DLL Storage, but I'm not sure what the best way to handle that set would be (and if there are others that could appear in the future).

rjmccall commented:
For type metadata, you need to make sure that tryEmitConstantTypeMetadataRef fails when the type metadata is defined externally.

For VWTs, we already talked about just emitting local copies of standard VWTs instead of trying to use copies in the stdlib.

Where do we emit constant references to protocol descriptors?  In protocol conformance records?

compnerd commented:
Sure, need to redo the build, but should be able to extract that

compnerd commented:
The current set  of imported symbols in the stdlib:

```
@"$SypN" = external dllimport global %swift.full_type
@"$SytN" = external dllimport global %swift.full_type
@"$SyXlN" = external dllimport global %swift.full_type
@"$SBoN" = external dllimport global %swift.full_type, align 8
@"$SBi64_N" = external dllimport global %swift.full_type, align 8
@"$SBi64_WV" = external dllimport global i8*, align 8
@"$SBoWV" = external dllimport global i8*, align 8
@"$SyycWV" = external dllimport global i8*, align 8
@"$SBi8_WV" = external dllimport global i8*, align 8
@"$SBpWV" = external dllimport global i8*, align 8
@"$SBbN" = external dllimport global %swift.full_type, align 8
@"$SBi8_N" = external dllimport global %swift.full_type, align 8
@"$SBi32_N" = external dllimport global %swift.full_type, align 8
@"$SBi16_N" = external dllimport global %swift.full_type, align 8
```

jckarter commented:
All of those symbols are defined in the C++ runtime. Since the standard library and runtime are always linked together, they shouldn't need dll storage at all when referenced from the standard library.

jckarter commented:
(sorry, clicked "close and comment" by accident)

compnerd commented:
@jckarter - I agree that they shouldn't need the imported DLL storage - thats what this PR is trying to alter.  The other ones are now taken care of.  The VWT case requires global emission with LinkOnceODR.  The few type metadata address point cases needs to be resolved.

I think that with the latter fixed, this can go in, the VWT emission is a larger change which currently I believe to be the last blocking thing to working on the tests (I have ~85% of the tests passing, there are a number of them which do not build due to the VWT for `Builtin.Object` failing to link).

I don't see the definition of these types in the runtime, but if they are defined in the runtime rather than the stdlib, then it sounds like a build issue that I should be able to fix quickly.

jckarter commented:
These metadata and value witness table symbols are all defined in KnownMetadata.cpp. You may need to change our code for emitting type metadata references for these types (empty tuple, Any, AnyObject, etc.) so that we go through the normal runtime entry points for them instead of assuming we can directly link them on Windows.

compnerd commented:
@jckarter - so, looking at this, the dllimport is definitely coming from the generation of the declaration in the compiler, not the runtime side.  Trying to figure out how to identify if the reference for the type metadata is being generated for building the runtime.  If we can fix that, then the Value Witness Tables can be handled in a subsequent change.

compnerd commented:
@jckarter @rjmccall - how bad of a solution is this?  It fixes all the instances of the incorrect linkage in the standard library build.  I feel that the approach in the IRLinkage computation is a bit heavy handed, but it is definitely far more expedient as I couldn't get it isolated enough to fix it at all the sites.

compnerd commented:
@swift-ci please test

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1822,8 +1822,8 @@ void irgen::updateLinkageForDefinition(IRGenModule &IGM,
-      getIRLinkage(linkInfo, entity.getLinkage(ForDefinition),
-                   ForDefinition, entity.isWeakImported(IGM.getSwiftModule()));
+      getIRLinkage(linkInfo, entity.getLinkage(ForDefinition), ForDefinition,
+                   entity.isWeakImported(IGM.getSwiftModule()));
```
This is no longer a real change; please revert it to make the patch simpler.

compnerd replied:
OK

rjmccall commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1850,10 +1850,12 @@ LinkInfo LinkInfo::get(const UniversalLinkageInfo &linkInfo,
                        ForDefinition_t isDefinition) {
   LinkInfo result;
+  ForDefinition_t ForDefinition =
+      ForDefinition_t(swiftModule->isStdlibModule() || isDefinition);
```
...please don't shadow `ForDefinition` here, it is exceptionally confusing.

Please check `isDefinition` first, and add a comment explaining what this is about.  I'm not sure I'm totally following how changing the SILLinkage of the entity gets you the right `dllimport`-ness.

compnerd replied:
Will do that.  So, its a pretty heavy hammer: basically, what I'm saying is that any declaration when building the standard library is for the definition.  This prevents it from being marked as an external symbol, which results in the `getIRLinkage` later not marking it as being DLL Import, which means that the DLL storage is now correct.  This happens to work because any declaration which is imported is either marked as such from the ClangImporter (it is from the shims) or is  `@_silgen_name`'ed , and will be marked for import.

rjmccall replied:
Okay.  I can live with that hack; please include that as a comment, though.  We should probably integrate these linkage calculations together in the long term.

compnerd commented on lib/IRGen/IRGen.cpp:
```diff
@@ -660,14 +660,23 @@ static void embedBitcode(llvm::Module *M, const IRGenOptions &Opts)
+  MDNode->addOperand(llvm::MDTuple::get(Context,
+                                        {llvm::MDString::get(Context,
+                                                             "standard-library"),
+                                         llvm::ConstantAsMetadata::get(Value)}));
```
TBH, its because I emulated what we do in clang for ObjC, I have no strong reason to model this this way or another way.  What would you prefer?  I didn't see a simple way to add a flag to the (LLVM) module btw.

rjmccall replied:
I think it's probably fine; if Clang does it that way now, there's probably a reason.

compnerd commented on lib/IRGen/IRGenModule.h:
```diff
@@ -1026,6 +1026,8 @@ class IRGenModule {
   llvm::Module *releaseModule();
   llvm::AttributeList getAllocAttrs();
 
+  bool IsStandardLibrary() const;
```
Sure, that's easy to fix.

compnerd commented:
@swift-ci please test

compnerd commented:
I believe I've addressed the feedback, and I verified this against the current top of tree.  The actual SR issue has been resolved with this change.  Thanks for the pointers @rjmccall and @jckarter !
